File: crashreporter/src/main/java/net/iamsilver/crashreporter/utils/CrashUtils.java
package net.iamsilver.crashreporter.utils;

import android.util.Log;

import net.iamsilver.crashreporter.CrashReporter;

import java.io.File;
import java.io.FilenameFilter;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;

public class CrashUtils {

    private static final String TAG = CrashUtils.class.getSimpleName();

    private static final FilenameFilter crashReportsFilter = new FilenameFilter() {
        @Override
        public boolean accept(File dir, String name) {
            return name.endsWith(Constants.CRASH_SUFFIX + Constants.FILE_EXTENSION);
        }
    };

    private static final FilenameFilter exceptionReportsFilter = new FilenameFilter() {
        @Override
        public boolean accept(File dir, String name) {
            return name.endsWith(Constants.EXCEPTION_SUFFIX + Constants.FILE_EXTENSION);
        }
    };

    private static final FilenameFilter allReportsFilter = new FilenameFilter() {
        @Override
        public boolean accept(File dir, String name) {
            return true;
        }
    };

    //

    private CrashUtils() {
        // this class is not publicly instantiable
    }

    //

    public static ArrayList<File> getReports() {
        return getReports(allReportsFilter);
    }

    public static ArrayList<File> getCrashReports() {
        return getReports(crashReportsFilter);
    }

    public static ArrayList<File> getExceptionReports() {
        return getReports(exceptionReportsFilter);
    }

    private static ArrayList<File> getReports(final FilenameFilter filter) {
        File[] files = new File(getCrashReportsPath()).listFiles(filter);
        return new ArrayList<>(Arrays.asList(files));
    }

    public static void saveCrashReport(final Throwable throwable) {
        saveReport(throwable, Constants.REPORT_TYPE_CRASH);
    }

    public static void saveExceptionReport(final Throwable throwable) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                saveReport(throwable, Constants.REPORT_TYPE_EXCEPTION);
            }
        }).start();
    }

    private static void saveReport(Throwable throwable, int reportType) {
        String suffix = "";
        if (reportType == Constants.REPORT_TYPE_CRASH) {
            suffix = Constants.CRASH_SUFFIX;
        } else if (reportType == Constants.REPORT_TYPE_EXCEPTION) {
            suffix = Constants.EXCEPTION_SUFFIX;
        }
        String filename = getCrashLogTime() + suffix + Constants.FILE_EXTENSION;
        FileUtils.writeToFile(filename, getStackTrace(throwable));
        NotificationUtils.showNotification(throwable.getLocalizedMessage(), reportType);
    }

    public static boolean deleteReport(File file) {
        if (!file.exists()) return true;

        boolean deleted = file.delete();
        if (!deleted) {
            Log.e(TAG, "Failed to delete report file: " + file.getAbsolutePath());
        }
        return deleted;
    }

    public static void clearCrashReports() {
        clearReports(crashReportsFilter);
    }

    public static void clearExceptionReports() {
        clearReports(exceptionReportsFilter);
    }

    public static void clearReports() {
        clearReports(new FilenameFilter() {
            @Override
            public boolean accept(File dir, String name) {
                return true;
            }
        });
    }

    private static void clearReports(final FilenameFilter filter) {
        File[] files = new File(getCrashReportsPath()).listFiles(filter);
        for (File file : files) {
            boolean deleted = file.delete();
            // TODO: ignored result of delete() call
        }
    }

    //

    private static String getCrashLogTime() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
        return dateFormat.format(new Date());
    }

    private static String getStackTrace(Throwable e) {
        final Writer result = new StringWriter();
        final PrintWriter printWriter = new PrintWriter(result);

        e.printStackTrace(printWriter);
        String crashLog = result.toString();
        printWriter.close();
        return crashLog;
    }

    public static String getCrashReportsPath() {
        File file = new File(CrashReporter.getApplicationContext().getExternalFilesDir(null), Constants.CRASH_REPORT_DIR);

        if (!file.exists()) {
            boolean created = file.mkdirs();
            if (!created) {
                // TODO: what happens if directory was not created?
                return null;
            }
        }
        return file.getAbsolutePath();
    }



}

File: crashreporter/src/main/kotlin/net/cebular/crashreporter/utils/CrashUtils.kt
package net.cebular.crashreporter.utils

import android.content.Context
import android.util.Log
import net.cebular.crashreporter.utils.Constants.LOG_TAG
import java.io.*
import java.text.SimpleDateFormat
import java.util.*

object CrashUtils {

    private val crashReportsFilter =
        FilenameFilter { _, name -> name.endsWith(Constants.CRASH_SUFFIX + Constants.FILE_EXTENSION) }

    private val exceptionReportsFilter =
        FilenameFilter { _, name -> name.endsWith(Constants.EXCEPTION_SUFFIX + Constants.FILE_EXTENSION) }

    private val allReportsFilter = FilenameFilter { _, _ -> true }

    //

    @JvmStatic
    fun getReports(context: Context): List<File> {
        return getReports(context, allReportsFilter)
    }

    @JvmStatic
    fun getCrashReports(context: Context): List<File> {
        return getReports(context, crashReportsFilter)
    }

    @JvmStatic
    fun getExceptionReports(context: Context): List<File> {
        return getReports(context, exceptionReportsFilter)
    }

    private fun getReports(context: Context, filter: FilenameFilter): List<File> {
        return getCrashReportsPath(context)?.let { File(it).listFiles(filter)?.toList() } ?: emptyList()
    }

    @JvmStatic
    fun saveCrashReport(context: Context, throwable: Throwable) {
        saveReport(context, throwable, Constants.REPORT_TYPE_CRASH)
    }

    @JvmStatic
    fun saveExceptionReport(context: Context, throwable: Throwable) {
        Thread { saveReport(context, throwable, Constants.REPORT_TYPE_EXCEPTION) }.start()
    }

    private fun saveReport(context: Context, throwable: Throwable, reportType: Int) {
        val suffix = when(reportType) {
            Constants.REPORT_TYPE_CRASH -> Constants.CRASH_SUFFIX
            Constants.REPORT_TYPE_EXCEPTION -> Constants.EXCEPTION_SUFFIX
            else -> ""
        }

        val filename = getCrashLogTime() + suffix + Constants.FILE_EXTENSION
        val reportFilePath = getCrashReportsPath(context) + File.separator + filename
        File(reportFilePath).writeText(getStackTrace(throwable))

        NotificationUtils.showNotification(context, throwable.localizedMessage, reportType)
    }

    @JvmStatic
    fun deleteReport(file: File): Boolean {
        if (!file.exists()) {
            return true
        }

        val deleted = file.delete()
        if (!deleted) {
            Log.e(LOG_TAG, "Could not delete crash report file '" + file.path + "'.")
        }
        return deleted
    }

    @JvmStatic
    fun clearCrashReports(context: Context) {
        clearReports(context, crashReportsFilter)
    }

    @JvmStatic
    fun clearExceptionReports(context: Context) {
        clearReports(context, exceptionReportsFilter)
    }

    @JvmStatic
    fun clearReports(context: Context) {
        clearReports(context, allReportsFilter)
    }

    private fun clearReports(context: Context, filter: FilenameFilter) {
        getCrashReportsPath(context)?.let {
            File(it).listFiles(filter)?.forEach { file ->
                val deleted = file.delete()
                if (!deleted) {
                    Log.e(LOG_TAG, "Could not delete crash report file '" + file.path + "'.")
                }
            }
        }
    }

    //

    private fun getCrashLogTime(): String {
        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
        return dateFormat.format(Date())
    }

    private fun getStackTrace(e: Throwable): String {
        val result: Writer = StringWriter()
        val printWriter = PrintWriter(result)

        e.printStackTrace(printWriter)
        val crashLog = result.toString()
        printWriter.close()
        return crashLog
    }

    @JvmStatic
    fun getCrashReportsPath(context: Context): String? {
        val file = File(
            context.getExternalFilesDir(null),
            Constants.CRASH_REPORT_DIR
        )
        if (!file.exists()) {
            val created = file.mkdirs()
            if (!created) {
                Log.e(LOG_TAG, "Could not create crash reports directory '" + file.path + "'.")
                return null
            }
        }
        return file.absolutePath
    }

}